newrelic/api/opentelemetry.py

```diff
@@ -0,0 +1,486 @@
+        else:
+            return bool(self.nr_transaction and (self.nr_transaction.sampled or (self.nr_transaction._sampled is None)))
+
+    def _is_remote(self):
+        # Remote span denotes if propagated from a remote parent
+        if self.otel_parent and self.otel_parent.is_remote:
+            return True
+        return False
+
+    def get_span_context(self):
+        if not (hasattr(self, "nr_trace") and self.nr_trace):
+            return otel_api_trace.INVALID_SPAN_CONTEXT
+
+        otel_tracestate_headers = None
+
+        return otel_api_trace.SpanContext(
+            trace_id=int(self.nr_transaction._trace_id, 16),
+            span_id=int(self.nr_trace.guid, 16),
+            is_remote=self._is_remote(),
+            trace_flags=otel_api_trace.TraceFlags(0x01 if self._is_sampled() else 0x00),
+            trace_state=otel_api_trace.TraceState(otel_tracestate_headers),
+        )
+
+    def set_attribute(self, key, value):
+        self.attributes[key] = value
+
+    def set_attributes(self, attributes):
+        for key, value in attributes.items():
+            self.set_attribute(key, value)
+
+    def _set_attributes_in_nr(self, otel_attributes=None):
+        if not (otel_attributes and hasattr(self, "nr_trace") and self.nr_trace):
+            return
+        for key, value in otel_attributes.items():
+            self.nr_trace.add_custom_attribute(key, value)
+
+    def add_event(self, name, attributes=None, timestamp=None):
+        # TODO: Not implemented yet.
+        # We can implement this as a log event
+        pass
+
+    def add_link(self, context=None, attributes=None):
+        # TODO: Not implemented yet.
+        pass
+
+    def update_name(self, name):
+        # Sentinel, MessageTrace, DatastoreTrace, and ExternalTrace
+        # types do not have a name attribute
+        self._name = name
+        if hasattr(self, "nr_trace") and hasattr(self.nr_trace, "name"):
+            self.nr_trace.name = self._name
+
+    def is_recording(self):
+        return self._is_sampled() and not (
+            hasattr(self, "nr_trace") and hasattr(self.nr_trace, "end_time") and self.nr_trace.end_time
+        )
+
+    def set_status(self, status, description=None):
+        # TODO: not implemented yet
+        pass
+
+    def record_exception(self, exception, attributes=None, timestamp=None, escaped=False):
+        if not hasattr(self, "nr_trace"):
+            if exception:
+                notice_error((type(exception), exception, exception.__traceback__))
+            else:
+                notice_error(sys.exc_info(), attributes=attributes)
+        else:
+            self.nr_trace.notice_error((type(exception), exception, exception.__traceback__), attributes=attributes)
+
+    def end(self, end_time=None, *args, **kwargs):
+        # We will ignore the end_time parameter and use NR's end_time
+
+        # Check to see if New Relic trace ever existed or,
+        # if it does, that trace has already ended
```

Comment:
Maybe store this value and simplify the logic rather than calling hasattr a bunch:
```
nr_trace = hasattr(self, "nr_trace", None)
if not nr_trace or nr_trace and getattr(nr_trace, "end_time", None):
    return